src/Common/FloatUtils.h

```diff
@@ -0,0 +1,49 @@
+#pragma once
+
+#include <base/DecomposedFloat.h>
+
+
+inline float convertFloat16ToFloat32(uint16_t float16_value)
+{
+    DecomposedFloat<Float16Tag> components(float16_value);
+
+    uint32_t old_sign = components.isNegative();
+    uint32_t old_exponent = components.exponent();
+    uint32_t old_mantissa = static_cast<uint32_t>(components.mantissa());
+
+    uint32_t new_exponent;
+    uint32_t new_mantissa;
+    uint32_t new_sign = old_sign << 31;
+
+    if (unlikely(old_exponent == 0x1F))
+    {
+        /// Inf, NaN
+        new_exponent = 0xFF;
```

Comment:
While looking at the failed tests, I've noticed two things:

- the output values became more accurate, because the old code introduced an extra rounding step, which technically alters the exact float32 value you would get from a pure half-to-float conversion. I think it much better now.
- the special values like `inf`, `nan` are not represented correctly, I believe this line causes this inconvenience. I hope that we need to do:

``` SQL
new_exponent = 0xFF << 23;
```

exponent takes bits [30..23]. So we must place that 0xFF at bits [30..23]-which means shifting it 23 bits left.
Current code: https://onlinegdb.com/EgElYyM1c
Code with shifting: https://onlinegdb.com/k2ynC3P9V

Reply:
Thanks, this is exactly the case. Fixed.